hooks/environment

```diff
@@ -53,11 +53,19 @@ function retry() {
   local -i max_attempts=$((retries + 1))
   local -i attempt_num=1
   local exit_code
-
-  until "$@" ; do
+  local stdin_value
+  set +e
+  read -sr -t0.01 stdin_value
```

Comment:
I’m looking into this timeout now, are there any concerns this could race for failure and introduce a source of flakey-ness?

Reply:
i don't think so the stdin will already be there, or will be not there ever (for the aws-cli command). i ran this fork on one of our projects with 200 parallel steps and it worked for all of them

Reply:
I delved in to the Bash source for this and found that if stdin is a regular file `read` disables the timeout behaviour. Bash uses a temporary file for herestrings so there shouldn’t be any timeout concerns for this.

I also found that `read -t0` can be used to essentially `select` on a file descriptor to find if there are any bytes ready to read.

```
$ bash --version
GNU bash, version 4.2.46(2)-release (x86_64-koji-linux-gnu)
Copyright (C) 2011 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
[ec2-user@ip-172-31-11-31 ~]$ bash -c 'ls -la /dev/fd/*; read -sr -t0 thing; echo $thing' <<< "foo"
ls: cannot access /dev/fd/3: No such file or directory
lr-x------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/0 -> /tmp/sh-thd-1632707943 (deleted)
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/1 -> /dev/pts/0
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/2 -> /dev/pts/0

[ec2-user@ip-172-31-11-31 ~]$ bash -c 'ls -la /dev/fd/*; read -sr -t0.01 thing; echo $thing' <<< "foo"
ls: cannot access /dev/fd/3: No such file or directory
lr-x------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/0 -> /tmp/sh-thd-1632717277 (deleted)
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/1 -> /dev/pts/0
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/2 -> /dev/pts/0
foo
[ec2-user@ip-172-31-11-31 ~]$ bash -c 'ls -la /dev/fd/*; read -sr -t1 thing; echo $thing' <<< "foo"
ls: cannot access /dev/fd/3: No such file or directory
lr-x------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/0 -> /tmp/sh-thd-1632727955 (deleted)
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/1 -> /dev/pts/0
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/2 -> /dev/pts/0
foo
[ec2-user@ip-172-31-11-31 ~]$ bash -c 'ls -la /dev/fd/*; read -sr -t1 thing; echo $thing'
ls: cannot access /dev/fd/3: No such file or directory
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/0 -> /dev/pts/0
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/1 -> /dev/pts/0
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:35 /dev/fd/2 -> /dev/pts/0
[1 second pause]
[ec2-user@ip-172-31-11-31 ~]$ bash -c 'ls -la /dev/fd/*; read -sr -t1 thing; echo $thing' </dev/null
ls: cannot access /dev/fd/3: No such file or directory
lr-x------ 1 ec2-user ec2-user 64 Sep 27 04:36 /dev/fd/0 -> /dev/null
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:36 /dev/fd/1 -> /dev/pts/0
lrwx------ 1 ec2-user ec2-user 64 Sep 27 04:36 /dev/fd/2 -> /dev/pts/0
```

> or will be not there ever (for the aws-cli command)

In light of the regular-file case findings it seems like this is being used as a fail fast for the no-input `aws get-login` case while supporting repeatable stdin for the `docker login` case. I wonder if there’s a neat way to make these cases explicit rather than implicit to save ourselves a comment on what this timeout value means to each case. What do you think?

Reply:
i removed the timeout, replaced it with the worlds laziest keyword option, what do you think?

Reply:
I think it looks great! Thanks for making that change :bow: